### PR TITLE
Feature/Optimize PageCounter Queries Release One [PLAT-1171]

### PR DESCRIPTION
**Pull request by pattisdr**
❗️ This is Part One of PageCounter Migration
[Part Two of Page Counter Migration here
](https://github.com/pattisdr/osf.io/pull/19) - will be released in a follow-up release.
## Purpose

Querying the PageCounter model is really inefficient.

## Changes
- Adds three migrations: 1) Schema migration - adds resource, file, action, and version fields. 2) Migration to add a multi-column index on the resource/file/action fields 3) Data migration to split the _id field into the resource, file, action, and version fields - if DEBUG_MODE is false, run the management command
- Adds temporary code to write to both the _id field, as well as the new resource/file/action/version fields.  PageCounter queries will continue to query the _id field until part II is merged.

## QA Notes
Would verify that viewing (if exposed in UI) and downloading files are still appropriately increasing counts.

## Side Effects

Will maybe slow down creating PageCounter objects temporarily until Part II merged.

## Ticket

https://openscience.atlassian.net/browse/PLAT-1171

### CODE REVIEW

File: osf/models/analytics.py [hunk 1]
```diff
@@ -55,17 +55,21 @@ def increment(cls, user_id, action, date_string):
...
     date = DateTimeAwareJSONField(default=dict)
 
     total = models.PositiveIntegerField(default=0)
     unique = models.PositiveIntegerField(default=0)
 
-    DOWNLOAD_ALL_VERSIONS_ID_PATTERN = r'^download:[^:]*:{1}[^:]*$'
+    action = models.CharField(max_length=128, null=False, blank=False, default='download', db_index=True)
+    guid = models.ForeignKey(Guid, related_name='pagecounters', null=True, blank=True, db_index=True)
+    file = models.ForeignKey('osf.BaseFileNode', null=True, blank=True, related_name='pagecounters', db_index=True)
+    version = models.IntegerField(null=True, blank=True, db_index=True)
```
Comment (pattisdr) [inline osf/models/analytics.py]: I split existing _id field of form action:node_id:file_id:version into four fields.  action = charfield, version is integerfield (if null, then this is the sum of all the versions) guid=FK and file= FK.

Made most null=True and blank=True.  Do we want to enforce that they're there?

File: osf/models/analytics.py [hunk 2]
```diff
@@ -55,17 +55,21 @@ def increment(cls, user_id, action, date_string):
...
+
+    class Meta:
+        index_together = (('action', 'guid', 'file', 'version'))
```
Comment (pattisdr) [inline osf/models/analytics.py]: Unsure about this...

File: osf/models/analytics.py [hunk 3]
```diff
@@ -55,17 +55,21 @@ def increment(cls, user_id, action, date_string):
         return True
 
 
-class PageCounter(BaseModel):
-    primary_identifier_name = '_id'
+class PageCounter(BaseModel, ObjectIDMixin):
+    _id = models.CharField(max_length=50, default=generate_object_id, unique=True, db_index=True)
```
Comment (pattisdr) [inline osf/models/analytics.py]: To not waste time writing the _id field as before...`action:node_id:file_id:version`, just inherited from ObjectIDmixin.  Left existing ids as is (which is why I override _id so I could bump up max length.  Do we expose this __id anywhere?  Is it even needed now?

File: osf/models/analytics.py [hunk 4]
```diff
@@ -55,17 +55,21 @@ def increment(cls, user_id, action, date_string):
...
-                           unique=True)  # 272 in prod
     date = DateTimeAwareJSONField(default=dict)
 
     total = models.PositiveIntegerField(default=0)
     unique = models.PositiveIntegerField(default=0)
 
-    DOWNLOAD_ALL_VERSIONS_ID_PATTERN = r'^download:[^:]*:{1}[^:]*$'
+    action = models.CharField(max_length=128, null=False, blank=False, default='download', db_index=True)
+    guid = models.ForeignKey(Guid, related_name='pagecounters', null=True, blank=True, db_index=True)
+    file = models.ForeignKey('osf.BaseFileNode', null=True, blank=True, related_name='pagecounters', db_index=True)
```
Comment (pattisdr) [inline osf/models/analytics.py]: Is file a protected variable name?

File: osf/migrations/0142_pagecounter_migration.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    four new columns: action(char), guid(fk), file(fk), version(int)
+    """
+    with connection.cursor() as cursor:
+        cursor.execute(
+            """
+            UPDATE osf_pagecounter PC
+            SET (action, guid_id, file_id, version) = (split_part(PC._id, ':', 1), G.id, F.id, NULLIF(split_part(PC._id, ':', 4), '')::int)
+            FROM osf_guid G, osf_basefilenode F
+            WHERE PC._id LIKE '%' || G._id || '%'
+            AND PC._id LIKE '%' || F._id || '%';
```
Comment (pattisdr) [inline osf/migrations/0142_pagecounter_migration.py]: Data migration to split every pagecounter _id into the action, guid, file, and version fields

File: osf/migrations/0143_pagecounter_schema.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+            model_name='pagecounter',
+            name='guid',
+            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, related_name='pagecounters', to='osf.Guid'),
+        ),
+        migrations.AddField(
+            model_name='pagecounter',
+            name='version',
+            field=models.IntegerField(blank=True, null=True, db_index=True),
+        ),
+    ]
```
Comment (pattisdr) [inline osf/migrations/0143_pagecounter_schema.py]: Schema migration - adds action, file, guid, and version fields.

action = charfield, version is integerfield (if null, then this is the sum of all the versions) guid=FK and file= FK.

Made most null=True and blank=True. Do we want to enforce that they're there?

File: osf/migrations/0144_pagecounter_index.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    dependencies = [
+        ('osf', '0143_pagecounter_schema'),
+    ]
+
+    operations = [
+        migrations.RunSQL([
+            'CREATE INDEX CONCURRENTLY page_counter_idx ON osf_pagecounter (action, guid_id, file_id);',
+        ], [
+            'DROP INDEX IF EXISTS page_counter_idx, RESTRICT;'
+        ])
```
Comment (pattisdr) [inline osf/migrations/0144_pagecounter_index.py]: Creates a multi-column index on action, guid, and file fields.  I chose these three instead of adding the version field to it, because this query is more common.   I can add an additional index on all four together if needed.

File: osf/migrations/0145_pagecounter_data.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    with connection.cursor() as cursor:
+        cursor.execute(
+            """
+            UPDATE osf_pagecounter PC
+            SET (action, guid_id, file_id, version) = (split_part(PC._id, ':', 1), G.id, F.id, NULLIF(split_part(PC._id, ':', 4), '')::int)
+            FROM osf_guid G, osf_basefilenode F
+            WHERE PC._id LIKE '%' || G._id || '%'
+            AND PC._id LIKE '%' || F._id || '%';
+            """
+        )
```
Comment (pattisdr) [inline osf/migrations/0145_pagecounter_data.py]: Data migration to split the _id field into guid, file, version, and action fields.

File: osf/models/analytics.py [hunk 9]
```diff
@@ -93,15 +99,34 @@ def clean_page(page):
...
         visited_by_date = session.data.get('visited_by_date', {'date': date_string, 'pages': []})
-
         with transaction.atomic():
-            model_instance, created = cls.objects.select_for_update().get_or_create(_id=cleaned_page)
+            # Temporary backwards compat - when creating new PageCounters, temporarily write to _id, guid, file, action, and version fields
+            try:
+                model_instance = cls.objects.get(_id=cleaned_page)
+            except cls.DoesNotExist:
+                guid, file, action, version = cls.deconstruct_id(cleaned_page)
+                model_instance = cls.objects.create(_id=cleaned_page, guid=guid, file=file, action=action, version=version)
```
Comment (pattisdr) [inline osf/models/analytics.py]: Temporarily writing to both the _id field, as well as guid/file/action/and version fields until Part II is merged.

Part II starts adding a long-form id for _id, instead of taking the time to construct the fancy _id that won't be necessary anymore now that we have the new columns to query on.

File: osf/migrations/0145_pagecounter_data.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('osf', '0144_pagecounter_index'),
+    ]
+
+    operations = [
+        migrations.RunPython(
+            separate_pagecounter_id, reverse_func
```
Comment (sloria) [inline osf/migrations/0145_pagecounter_data.py]: Drive-by it: Not sure if it makes a difference, but you can use the `RunSQL` operation here.

File: osf/models/analytics.py [hunk 11]
```diff
@@ -65,6 +66,11 @@ class PageCounter(BaseModel):
     total = models.PositiveIntegerField(default=0)
     unique = models.PositiveIntegerField(default=0)
 
+    action = models.CharField(max_length=128, null=False, blank=False, default='download', db_index=True)
+    guid = models.ForeignKey(Guid, related_name='pagecounters', null=True, blank=True, db_index=True)
```
Comment (erinspace) [inline osf/models/analytics.py]: I am wondering about using `guid` as the field name! Like perhaps it could be more descriptive in its role, like in the `Comment` model the `fk` to `Guid` is named `target` (and the other `root_target`) -- though maybe that's more confusing.  Like `Guid` is really more of a through table to other things that the `PageCounter` is connected to, right? I think I am overthinking this as I can't really think of a better alternative - don't like `target` cause we use that for so many other things... I dunno, what are your thoughts??

**Comment by pattisdr** [comment]
awesome, thanks for the review @erinspace

File: osf/models/analytics.py (hunk 11 above)
Comment (pattisdr) [reply osf/models/analytics.py]: I see what you're saying! I'm thinking about `resource`.  I used this where generalizing with NPD, methods that needed to work for both nodes and preprints, so I called it `resource`.  I think I got it from the [JSONAPI spec](https://jsonapi.org/format/#fetching-resources)

File: osf/migrations/0144_pagecounter_index.py (hunk 7 above)
Comment (pattisdr) [reply osf/migrations/0144_pagecounter_index.py]: Thinking about it more, I've removed the individual indices off of those fields and added the `version` field to the index.

File: osf/migrations/0143_pagecounter_schema.py (hunk 6 above)
Comment (pattisdr) [reply osf/migrations/0143_pagecounter_schema.py]: Makes sense! I made "resource" formerly guid and file fields required in Part II.

File: osf/migrations/0155_pagecounter_data.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+        id_array = pc._id.split(':')
+
+        pc.action = id_array[0]
+        pc.resource_id = Guid.objects.get(_id=id_array[1]).id
+        pc.file_id = BaseFileNode.objects.get(_id=id_array[2]).id
+        pc.version = id_array[3] if len(id_array) == 4 else None
+        batch.append(pc)
+        progress_bar.update(i)
+    bulk_update(batch, update_fields=['action', 'resource_id', 'file_id', 'version'], batch_size=10000)
+    progress_bar.finish()
```
Comment (pattisdr) [inline osf/migrations/0155_pagecounter_data.py]: Was formerly in SQL but I ran it on Steve's new post-NPD staging dump and it's realllyy slow.  I modified to be more similar to Matt's pagecounter [NPD migration fix ](https://github.com/CenterForOpenScience/osf.io/pull/8841)

SQL here
```SQL
def separate_pagecounter_id(state, schema):
    """
    Splits the data in pagecounter _id field of form action:guid_id:file_id:version into
    four new columns: action(char), guid(fk), file(fk), version(int)
    """
    with connection.cursor() as cursor:
        cursor.execute(
            """
            UPDATE osf_pagecounter PC
            SET (action, guid_id, file_id, version) = (split_part(PC._id, ':', 1), G.id, F.id, NULLIF(split_part(PC._id, ':', 4), '')::int)
            FROM osf_guid G, osf_basefilenode F
            WHERE PC._id LIKE '%' || G._id || '%'
            AND PC._id LIKE '%' || F._id || '%';
            """
        )
```

Comment (brianjgeiger) [reply osf/migrations/0155_pagecounter_data.py]: I think the only flaw with the SQL above is trying to compare to the `PC._id` in the WHERE clauses rather than the other way around, because any time you have to search on a field where the `%` is not at the end, it will have to go through the records by hand rather than using the index. 

So the following _should_ be much more performant than the previous SQL and, hopefully, the python loops. I'm testing now.

```
UPDATE osf_pagecounter PC
SET
    action = split_part(PC._id, ':', 1),
    resource_id = Guid.id,
    file_id = File.id,
    version = NULLIF(split_part(PC._id, ':', 4), '')::int
FROM osf_guid Guid, osf_basefilenode File
WHERE Guid._id = split_part(PC._id, ':', 2)
AND File._id = split_part(PC._id, ':', 3);
```

**Comment by pattisdr** [comment]
note to self: after merging in Brian's work, change action so it is nullable (and get rid of `download` as default. (will update in part II to revert this).  Also rebase migrations before we get this on staging.
